src/Microsoft.AspNetCore.Sockets.Common/Internal/Formatters/ServerSentEventsMessageParser.cs

```diff
@@ -0,0 +1,208 @@
+    {
+        const byte ByteCR = (byte)'\r';
+        const byte ByteLF = (byte)'\n';
+        const byte ByteSpace = (byte)' ';
+
+        private InternalParseState _internalParserState = InternalParseState.ReadMessageType;
+        private IList<byte[]> _data = new List<byte[]>();
+
+        public ParseResult ParseMessage(ReadableBuffer buffer, out ReadCursor consumed, out ReadCursor examined, out Message message)
+        {
+            consumed = buffer.Start;
+            examined = buffer.End;
+            message = new Message();
+            var messageType = MessageType.Text;
+            var reader = new ReadableBufferReader(buffer);
+
+            var start = consumed;
+            var end = examined;
+
+            while (!reader.End)
+            {
+                if (ReadCursorOperations.Seek(start, end, out var lineEnd, ByteLF) == -1)
+                {
+                    if(end != buffer.End)
+                    {
+                        return ParseResult.Incomplete;
+                    }
+                    throw new FormatException("Expected a '\r\n' line ending");
+                }
+
+                lineEnd = buffer.Move(lineEnd, 1);
+                var line = ConvertBufferToSpan(buffer.Slice(start, lineEnd));
+                reader.Skip(line.Length);
+
+                if(!(line.Length > 1))
+                {
+                    throw new FormatException("Error parsing data from event stream");
+                }
+
+                //Strip the \r\n from the span
+                line = line.Slice(0, line.Length - 2);
+
+                switch (_internalParserState)
+                {
+                    case InternalParseState.ReadMessageType:
+                        messageType = GetMessageType(line);
+                        start = lineEnd;
+                        _internalParserState = InternalParseState.ReadMessagePayload;
+                        consumed = lineEnd;
+
+                        //Peek into next byte. If it is a carriage return byte, then advance to the next state
+                        if (reader.Peek() == ByteCR)
+                        {
+                            _internalParserState = InternalParseState.ReadEndOfMessage;
+                        }
+                        break;
+                    case InternalParseState.ReadMessagePayload:
+                        if (!StartsWithDataPrefix(line))
+                        {
+                            throw new FormatException("Expected the message prefix 'data: '");
+                        }
+
+                        //Slice away the 'data: '
+                        var newData = line.Slice(line.IndexOf(ByteSpace) + 1).ToArray();
+                        start = lineEnd;
+                        _data.Add(newData);
+
+                        //Peek into next byte. If it is a carriage return byte, then advance to the next state
+                        if (reader.Peek() == ByteCR)
```

Comment:
Since both cases which exit the `switch` via `break` need to do this check, let's move it outside the `switch` entirely. The last case exits with `return` so it will skip it anyway.